### ISSUE TITLE
Text with ruby gets shifted left

### ISSUE THREAD

**Issue by MyK00L** [koreader/koreader#11021]
* KOReader version: 2023.08
* Device: Linux/Android

#### Issue
Rarely text with ruby appears shifted to the left, often overlapping with previous text and leaving empty space to the right where it should have been.
I don't know much of how this works, so I'm not sure if this is a bug with koreader or the underlying rederer...

#### Steps to reproduce
If you open a long enough epub file with japanese text containing ruby, you should see it happen somewhere.
For the example image below, I've taken a [random txt from aozora bunko](https://www.aozora.gr.jp/cards/000160/card3354.html) and converted it to epub with [this tool](https://github.com/sgqy/AozoraEpub3).
I haven't been able to identify anything particular that might cause this bug, but:
- the text appears in the right spot during partial rendering
- some settings, like font kerning method, don't seem to affect this
- other settings (eg. "Two Columns") seems to affect wether it happens or not at particular spots (but it doesn't happen in a spot it could happen somewhere else)
- I haven't found a setting that gets rid of this bug entirely.

##### example
![20231018_12h02m03s_grim](https://github.com/koreader/koreader/assets/22148361/a744a894-d18b-456e-878f-2a757c93f02b)
Here, near the center of the left page in "かたまりと化している。", "化" is moved to the left.

##### log
This is the log i got in stdout, opening koreader and closing after the page was rendered.
[log.txt](https://github.com/koreader/koreader/files/13019262/log.txt)

**Comment by poire-z** [comment]
A bit confusing what you explain, but I guess I got it: it's not really the ruby text itself which is shifted to the left, but the base text on which the ruby text sits, right ?
![image](https://github.com/koreader/koreader/assets/24273478/b96fc22b-ece4-4a8c-909a-cb1dd118d137)
It feels like there may be some space at the end and included in the base text, which may be is accounted in the width of the base text (in my green rectangle), and the ruby text is properly centered above it. Does it feel that way ? (I don't read Japanese/Chinese., I can't say if the space is expected or bogus).

If that's the case, I dunno if we should have ignored that whitespace char (the Ruby rendering rules are complicated, don't want to re-read them :) and it may not be really easy to handle that).

Although it doesn't look like it in your source HTML file:
![image](https://github.com/koreader/koreader/assets/24273478/623b9d9d-a500-4c9d-9dcc-cffbc5ff01d6)

And if I download the HTML and just have KOReader read it:

![image](https://github.com/koreader/koreader/assets/24273478/0901cc2d-c942-4177-8140-217699ca8e22)

The `View HTML` looks fine, without any extra spaces:
![image](https://github.com/koreader/koreader/assets/24273478/1a0c203a-d199-48d6-8926-b3eda4351753)

It feels like your 化 is drawn over the preceding と ? Is that it ? (It's hard for me to notice that the resulting ![image](https://github.com/koreader/koreader/assets/24273478/447d1364-80ab-4af1-bbc2-0e15498fe865) is not a valid Japanese character :)

Just trying to understand what is really happening and where is the problem.
If I enable `Style tweak > Text alignment > Left align most text` (so, no text justification of the paragraph), it looks ok ?
![image](https://github.com/koreader/koreader/assets/24273478/76dc4d83-a4eb-4a75-ace6-f8689d70a1ce)

Then, it might be an issue with text justification and inline tables (that we use for Ruby and MathML stuff.... I remember I had a hard time with that :/)

With a book style tweak containing `ruby { border: 1px solid red }`:

`[x] Left alignt most text`:
![image](https://github.com/koreader/koreader/assets/24273478/8469796a-902e-4242-bc30-95a4bbe043bc)

`[ ] Left alignt most text`:
![image](https://github.com/koreader/koreader/assets/24273478/1d77d734-e1b9-4b5c-8027-88499b20afa8)

It even happens if I remove the hidden/invisible `<rp>`.
No real idea what's happening :/
I'll investigate when I have the time - please tell if you agree or not with my observations/guesses.

**Comment by poire-z** [comment]
Strangely, it doesn't happen if I make that line longer (so it is NOT the last non-justified line of the paragraph), it looks ok:
![image](https://github.com/koreader/koreader/assets/24273478/d571c6aa-b8a0-46e4-b687-88f53a85818b)

(So, may be just some "stuff done when paragraph/line is justified" that should not be done on the last line.)

**Comment by MyK00L** [comment]
Wow, thanks, that was fast, now I can disable justification and be happy.
If it can help, here's a simple example:
```html
<!DOCTYPE html>
<html>
<head>
<style>
ruby { border: 1px solid black }
rt { border: 1px solid brown }
.j1 {text-align: justify;}
.a {color: blue;}
.b {color: red;}
</style>
</head>
<body>
<p class="j1 a">可可可可可可可可可可可可可可可可可可可可可可可可可可可可可可可可可可可可可可可可可<ruby>可<rt>rt</rt></ruby></p>
<p class="j1 b">可可可可可可可可可可可可可可可可可可可可可可可可可可可可可可可可可可<ruby>可<rt>rt</rt></ruby>可可可可可可可</p>
<p class="j1 a">可可可可可可可可可可可可可可可可可可可可可可可可可<ruby>可<rt>rt</rt></ruby>可可<ruby>可<rt>rt</rt></ruby>可可<ruby>可<rt>rt</rt></ruby>可可<ruby>可<rt>rt</rt></ruby>可可可可可可可</p>
<p class="j1 b">可可可可可可可可可可可可可可可可可可可可可可<ruby>可<rt>rt</rt></ruby>可可<ruby>可<rt>rt</rt></ruby>可可<ruby>可<rt>rt</rt></ruby>可可<ruby>可<rt>rt</rt></ruby>可可<ruby>可<rt>rt</rt></ruby>可可可可可可可</p>
<p class="j1 a">可可<ruby>可<rt>rt</rt></ruby>可可可可可可可可可可可可可可可可可可可可可可可可可可可可可可可<ruby>可<rt>rt</rt></ruby>可可可可可可可</p>
<p class="j1 b">可可可可可可可可可可可可可可可可可可可可可可可可可可可可可可可可可可<ruby>可<rt>rt</rt></ruby>可可可可可可可可可</p>
<p class="j1 a">可可可可可可可可可可可可可可可可可可可可可可可可可可可可aaaaaaaaaaa<ruby>bbbbbbbbbbbbb<rt>rt</rt></ruby></p>
</body>
</html>
```
This is how it's rendered:
![20231018_20h50m55s_grim](https://github.com/koreader/koreader/assets/22148361/a07b2b45-ffd6-4aac-9840-7c5d92308693)
Some points about the above example:
- the reason only the last lines are affected might be because there is no justification to be done in full lines, since alle characters have the same width(?)
- characters further to the right get shifted more
- the `<ruby>` "box" stays in the right place but the `<rt>` box moves
- having a ruby on the first line somehow makes the ruby on the second line to have no problem
- I could not reproduce it with english characters only

**Comment by poire-z** [comment]
> the reason only the last lines are affected might be because there is no justification to be done in full lines, since alle characters have the same width(?)

There may be justification happening, when the paragraph width is not a multiple of the (fixed) CJK char width: we will add space in-between each char so the right most ones reach the right edge of the paragraph width.
In case you didn't know, there's a style tweak under `Paragraph>` that achieves what you expect by actually reducing the paragraph width, so it may also fix your problem:
![image](https://github.com/koreader/koreader/assets/24273478/d7a841b5-0da6-4418-b41a-ec76c6d3e2ef)

Anyway, the code that deal with the last line was added by https://github.com/koreader/crengine/commit/c7c6bef9.
I thought the issue was here, but actually, not really: it is due to some optimisation we do, and that affects what that last line handling does:

- when rendering paragraph for drawing the current page, we do: 1) (processParagraph) break the long html `<p>` content to make lines, cutting it when it becomes too long 2) (addLine) dealing with each line to make "words" 3) (alignLine) to reposition the "words" on the line according to text-align CSS and a few other tweaks, including the above commit. When this is done, the resulting object is cached in memory, for re-use if we redraw that page.
 
- when rendering the full book, we need to render each paragraph, to get each line height and be able to compute that paragraph full height. This is optimized, with a cache size of 1 (as we will render so many paragraphs that they would not live long in the cache) - but also by not applying the step "3) alignLine)" mentionned above, as it's not needed to get the line height.

- but that "3) alignLine)" step can't be skipped when the line contains "inlineBox" (inline-block, inline-table, ruby, mathml) as we need to render each of them to get their height (that participate in the resulting line height) - so we do that step for each line that contains a ruby. When rendering such inlineBox, its x/y/w/h is hard-stored in the document cache (not the volatile cache mentionned above), and we don't re-render them when we meet them again when we reprocess that paragraph for drawing, as it may be expensive, and there's no reason for these x/y/w/h to change.

Because of all that, what is happening with the above commit is that when we process a paragraph with a first line without ruby is that we don't alignLine(), and we don't compute the added spacing to each char to have that line justified: the m_cjk_prev_line_added_space_div/mod stays 0.
When we render the last line, with some ruby, we do go thru alignLine, and use 0 for the char spacing, and we position the ruby box there as if it doesn't need any repositionning.

Later, when we are drawing that paragraph, we go thru that, but for all lines, so we may get m_cjk_prev_line_added_space_div/mod non-0 this time, and we will move each char of that last line by the spacing we noticed and remember on the previous line. And all is well.
Except that for these inlineBox ruby, that will be drawn at the position we stored the first time, so as if all previous characters have not moved to the right, and they will overlap with their previous character that were indeed correctly moved a little bit to the right to align with the previous line.

This explains everything that you and I have noticed above :) (except may be `the <ruby> "box" stays in the right place but the <rt> box moves`, but may be it's just the drawing of that box that is done by the paragraph code itself and not by the inner inlineBox content...)

TL;DR: phew, it's not really a ruby/table computing/drawing issue: just some optimization I'll have to unoptimize in some cases (no idea yet what's the best way to go at that).
